### PR TITLE
Workload configuration

### PR DESCRIPTION
**Pull request by csrwng**
This PR introduces a new CRD to allow fine grained configuration of a
HostedControlPlane's workload. It allows configuring settings that
affect scheduling such as affinity, tolerations and priority class as
well as resource requests, security context, etc.

To enable this configuration, a WorkloadConfiguration CR must be created
in the same namespace as a HostedControlPlane resource. Only one
WorkloadConfiguration resource may be present in a namespace. More than
one resource will result in reconciliation errors and will prevent that
particular control plane from being reconciled.

### CODE REVIEW

File: api/v1alpha1/workloadconfig.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func init() {
+	SchemeBuilder.Register(&WorkloadConfiguration{})
+	SchemeBuilder.Register(&WorkloadConfigurationList{})
+}
+
+// WorkloadConfiguration allows additional tuning of control plane workloads
```
Comment (enxebre) [inline api/v1alpha1/workloadconfig.go:13]: may be it's just me but "workloadconfig" makes me think of config for workloads running on the guest cluster.

**Comment by enxebre** [comment]
>To enable this configuration, a WorkloadConfiguration CR must be created
in the same namespace as a HostedControlPlane resource. Only one
WorkloadConfiguration resource may be present in a namespace. More than
one resource will result in reconciliation errors and will prevent that
particular control plane from being reconciled.

Are these assumptions meant to be documented?
Does this mean that the HCP namespace needs to be created ahead of time?
What's the reasoning behind that vs having an optional ref to this config in the HC CR?

File: api/v1alpha1/workloadconfig.go (hunk 1 above)
Comment (sjenning) [reply api/v1alpha1/workloadconfig.go:13]: I also think this.  Workload is extremely generic.  We are configuring a cluster.  I like `HostedClusterConfiguration` personally.

File: api/v1alpha1/workloadconfig.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+	// Annotations that should be applied to all deployments in the control plane
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Deployments contains configuration details for control plane component deployments
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	Deployments []DeploymentWorkloadConfiguration `json:"deployments,omitempty"`
```
Comment (sjenning) [inline api/v1alpha1/workloadconfig.go:47]: How will people know what the names of the deployments are without looking through the code?

**Comment by sjenning** [comment]
We probably talked about this but since there is a 1:1 relationship with the HCP resource, but no explicity binding between the two, only implicit in the resource being present in the HCP namespace,  why do we not put all this in say a `spec.deploymentConfiguration` on the HCP spec?

**Comment by ironcladlou** [comment]
I think this entire API is too generic and leaks implementation details out into the public API (deployment and container names, etc.) It also seems to introduce something like a fractal explosion of the test matrix that we'll never sufficiently cover.

Affinity rules can easily lead to breakages that I see no strategy for detecting or reporting. Letting the user override readiness and liveness probes seems like a recipe for users breaking components the operator is supposed to be rolling out in ways we've tested and controlled for.

What are the examples of underlying requirements driving this?

**Comment by csrwng** [comment]
@ironcladlou You can take a look at the scheduling parts in ROKS manifests:
https://github.com/openshift/ibm-roks-toolkit/blob/master/assets/kube-apiserver/kube-apiserver-deployment.yaml#L17-L56

**Comment by ironcladlou** [comment]
> @ironcladlou You can take a look at the scheduling parts in ROKS manifests:
> https://github.com/openshift/ibm-roks-toolkit/blob/master/assets/kube-apiserver/kube-apiserver-deployment.yaml#L17-L56

This shows what somebody _is_ doing, but not _why_, and doesn't speak to whether we should repeat it here, or as proposed...

**Comment by csrwng** [comment]
@ironcladlou the IBM team can better explain why they're using these settings, but it has to do with how they choose to schedule control plane workloads on their tugboat clusters. One of the reasons for making this configurable is that it is likely that every provider will want to customize the scheduling of their control plane workloads.

**Comment by ironcladlou** [comment]
> @ironcladlou the IBM team can better explain why they're using these settings, but it has to do with how they choose to schedule control plane workloads on their tugboat clusters. One of the reasons for making this configurable is that it is likely that every provider will want to customize the scheduling of their control plane workloads.

Even if we grant at least one user wants some manner of scheduling control of a component or set of components (setting aside whether it's really likely other providers need the same thing), is this API the way to achieve it?

For example, how does letting users override readiness probes for arbitrary containers in arbitrary deployments managed by HyperShift enable users to solve whatever scheduling challenge you're describing?

I just can't yet see a line from "IBM wants some scheduling control" to this API.

**Comment by relyt0925** [comment]
Let me provide reasoning

**Comment by relyt0925** [comment]
There's a few key reasons for what we do:

1) Multi zone deployments: that's probably the most simple (critical to our SLA and offering)
2) Tolerations: We taint our management cluster nodes to ensure only specific types of workload can run on specific types of machines. It helps to isolate customer master workload from other random components that might run in the management cluster that could cause disruptions. We only want master workload on nodes we use for masters
3) Restart Date: There is a want by users to be able to restart components without having to do a explicit patch update (helps sometimes for long running components, recovering from a memory leak or after overloading the apiserver, bad webhook, etc). Restart date gives a way for a customer to trigger rolling restarts of  all components
4) Pod affinity based on clusterid: We want to keep master components of a given cluster close to each other whenever possible. Significantly helps with webhook traffic and helps keep a given control plane confined to a node. We have seen situations with customers that run very busy and high scale masters (lots of api requests, etc) if their components are spread across a bunch of nodes there can be resource starvation across a large percentage of the management cluster fleet. We want to avoid that since that can cause evictions of other customer workload. Therefore: we aim to keep a given master's components in a given zone colocated on a node

**Comment by relyt0925** [comment]
These are all things that won't ultimately be IBM specific and be experienced by any team using hypershift at a large scale. This is some of the general operational tricks we have picked up during operating the offering.

**Comment by ironcladlou** [comment]
Thanks for the detailed requirements, that's very useful.

> There's a few key reasons for what we do:
> 
> 1. Multi zone deployments: that's probably the most simple (critical to our SLA and offering)

Can you elaborate further on this? Which specific components within a single hostedcluster are you referring to, and how would you expect multi zone to be achieved?

> 2. Tolerations: We taint our management cluster nodes to ensure only specific types of workload can run on specific types of machines. It helps to isolate customer master workload from other random components that might run in the management cluster that could cause disruptions. We only want master workload on nodes we use for masters

This makes sense, but is the implication that every managed pod related to the hosted control plane should be taken together in this regard with tolerations applied to all the pods? Or is some finer level of control over individual components of the control plane required (e.g. a difference in tolerations between the apiserver and, say, the KCM or etcd?)

> 3. Restart Date: There is a want by users to be able to restart components without having to do a explicit patch update (helps sometimes for long running components, recovering from a memory leak or after overloading the apiserver, bad webhook, etc). Restart date gives a way for a customer to trigger rolling restarts of  all components

This sounds like bugs that need described and resolved (e.g. bugs in liveness or readiness checking). If there are specific reproducers here, that would be useful. I'm not sure this represents an API need as described.

> 4. Pod affinity based on clusterid: We want to keep master components of a given cluster close to each other whenever possible. Significantly helps with webhook traffic and helps keep a given control plane confined to a node. We have seen situations with customers that run very busy and high scale masters (lots of api requests, etc) if their components are spread across a bunch of nodes there can be resource starvation across a large percentage of the management cluster fleet. We want to avoid that since that can cause evictions of other customer workload. Therefore: we aim to keep a given master's components in a given zone colocated on a node

As with the tolerations requirement, does this apply uniformly to every pod for a hosted control plane? If not, then at what granularity or grouping of components would they apply? I also wonder if this represents a need for more reasonable defaulting on the hypershift side. Is there any default affinity rules for components which would obviate the need for API?

One more general response I have is that I consider every requirement above to be orthogonal. I do understand how the generic/catch-all configuration surface in this PR can provide a means to achieve them all one way or another, but I think to arrive at something supportable and coherent we should be discussing each piece separately and with more consideration for the support and overall API implications.

The API as described here isn't far from just turning off the operators and letting the user directly manipulate resources arbitrarily, which is a path I doubt we want to go down. I suggest we'll have better luck finding the right APIs if we separate the requirements discussions out a little.

**Comment by ironcladlou** [comment]
We had an offline discussion about these topics and identified several actionable approaches and followup discussions, all of which taken together seem to eliminate the need for the API as proposed here. Let's pause this PR and break out the discussions into issues describing our intended direction for each piece.

/hold

**Pull request closed without merging**